### ISSUE TITLE
about (compute_contra_memobank_loss)

### ISSUE THREAD

**Issue by DISAPPEARED13** [Haochen-Wang409/U2PL#27]
感谢各位的文章，这很有启发性！
我想请问下，关于调用loss_helper.compute_contra_memobank_loss中输入的参数low_mask和high_mask，
用法如下：
`
    low_valid_pixel = torch.cat((label_l, label_u), dim=0) * low_mask
`
`
    high_valid_pixel = torch.cat((label_l, label_u), dim=0) * high_mask
`

但是，这个mask是怎么定义呢？我留意到在train_semi line419中定义了
`                    low_mask_all = torch.cat(
                        (
                            (label_l.unsqueeze(1) != 255).float(),
                            low_entropy_mask.unsqueeze(1),
                        )
                    )
`
`
                        high_mask_all = torch.cat(
                            (
                                (label_l.unsqueeze(1) != 255).float(),
                                high_entropy_mask.unsqueeze(1),
                            )
                        )
`
这边是为了达到什么效果呢？

感谢百忙之中抽空看我的问题！

**Comment by Haochen-Wang409** [comment]
感谢您对我们工作的认可！

low_mask 为 1 的 pixel 代表是**低熵**的，即 reliable 的。而 high_mask 为 1 的 pixel 则相反，为**高熵**的，unreliable 的。
Eq. (4) 中的 z (anchor) 来自 student，z+ (positive samples) 和 z- (negative samples) 均来自于 teacher

**Comment by DISAPPEARED13** [comment]
感谢您的答疑！
我想再请问一下关于loss_helper中LINE 135中`negative_mask = rep_mask_high_entropy * class_mask`。

其中`rep_mask_high_entropy`表示根据alpha_t挑选高熵的像素，其中我根据我的数据集修改了low rank和high rank。
而`class_mask`为满足Eq.16的`negative samples`的二值表示。

但是经过乘之后得到全为`False`的类型为`bool`的`negative mask`，这种情况下导致key赋值为空数组。我理解全False的`negative mask`的情况下可能意思为不存在【难以区分】或者【不属于c】的设置。
这是在训练过程中出现，是正常的吗？或者您有没有遇到过这样的情况？

再次感谢您！

**Comment by Haochen-Wang409** [comment]
照理来说不应该出现这样的情况。是否是您修改的 ```low_rank``` 大于了本身数据集的类别个数？

**Comment by DISAPPEARED13** [comment]
感谢您的提醒，您确实是对的！
我在全监督的情况下将背景也作为一个类进行dice的计算，我疏忽了忘了改过来了。
关于const loss memory bank我还有一个问题叨扰下您：
我在训练的时候（目前迭代次数不多）发现，有时候`const_loss`为0.0。
我留意到对于如果没有满足` if ( len(seg_feat_low_entropy_list[i]) > 0 and memobank[valid_classes[i]][0].shape[0] > 0 ):`（见loss_helper.py LINE约169）的情况下，此处的const loss是置为0计算的。也就是说，会在大约初始若干次迭代中，const loss可能无法作用到模型的更新中，是如此吗？

再次感谢您！

**Comment by Haochen-Wang409** [comment]
是的，您的理解完全正确。
其中，```len(seg_feat_low_entropy_list[i]) > 0```用于判断是否有符合条件的 anchor，而```memobank[valid_classes[i]][0].shape[0] > 0```用于判断当前类别是否有负样本。
这两者在 contra_loss 的计算当中不可或缺。

**Comment by DISAPPEARED13** [comment]
您好！又叨扰您了。：）
我发现在训练的过程中，```self.alpha_t```和```self.delta_p```在筛选低熵像素集合中起到了决定性作用。根据论文，我发现alpha_0初始值设定为0.2，此后根据当前epoch进行灵活变动。而delta_p则是固定为0.3。但是对于我的任务来说，不知道是不是因为这个值设定不合适我的数据，导致```seg_feat_low_entropy_list```经常为空集。从而导致预测分割DICE全为0。
可以麻烦您说一下是怎么选取这两个值吗？

**Comment by Haochen-Wang409** [comment]
```alpha_t```我们做了消融实验，请参照 Tab. 7
```delta_p```我们参照了 [ReCo](https://github.com/lorenmt/reco).
